ext/pbr_ext/read.cpp

```diff
@@ -0,0 +1,172 @@
+#include <iostream>
+
+#include "read.h"
+#include "read_write.h"
+#include "encode.h"
+
+using namespace std;
+
+void ss_skip(ss_t& ss, int32_t n);
+void skip(ss_t& ss, wire_t wire_type);
+
+// conventional variables used in this file
+// ss_t& ss  - the read buffer
+// Msg& msg  - the metamessage
+// Fld& fld  - the metafield
+// VALUE obj - the object being written
+// VALUE val - the field value on obj being written
+
+// these macros are unhygienic, but that's ok since they are
+// local to this file.
+#define DEF_RV(type)  VALUE rv_##type(ss_t& ss, Msg& msg, Fld& fld, vector<LazyField> *lazy_fields)
+#define INFLATE(val)  RTEST(fld.inflate) ? rb_funcall(fld.inflate, ID_CALL, 1, (val)) : (val)
+
+DEF_RV(INT32)    { return (INT2NUM          (r_varint32(ss)));  }
+DEF_RV(UINT32)   { return (UINT2NUM         (r_varint32(ss)));  }
+DEF_RV(INT64)    { return (LL2NUM           (r_varint64(ss)));  }
+DEF_RV(UINT64)   { return (ULL2NUM          (r_varint64(ss)));  }
+DEF_RV(SINT32)   { return (INT2NUM (zz_dec32(r_varint32(ss)))); }
+DEF_RV(SINT64)   { return (LL2NUM  (zz_dec64(r_varint64(ss)))); }
+DEF_RV(SFIXED32) { return (INT2NUM          (r_int32   (ss)));  }
+DEF_RV(FIXED32)  { return (UINT2NUM         (r_int32   (ss)));  }
+DEF_RV(SFIXED64) { return (LL2NUM           (r_int64   (ss)));  }
+DEF_RV(FIXED64)  { return (ULL2NUM          (r_int64   (ss)));  }
+
+DEF_RV(ENUM) {
+  VALUE val = INT2NUM(r_varint32(ss)); 
+  if (msg.model->validate_on_read)
+    validate_enum(msg, fld, val);
+  return val;
+}
+
+DEF_RV(FLOAT) {
+  uint32_t v = r_int32(ss);
+  return (DBL2NUM((double)REINTERPRET(float, v)));
+}
+
+DEF_RV(DOUBLE) {
+  uint64_t v = r_int64(ss);
+  return (DBL2NUM(REINTERPRET(double, v)));
+}
+
+DEF_RV(BOOL) {
+  uint32_t v = r_varint32(ss);
+  return (v == 0 ? Qfalse : Qtrue);
+}
+
+DEF_RV(STRING) {
+  int32_t len = r_varint32(ss);
+  VALUE rstr = rb_str_new(ss_read_chars(ss, len), len);
+  return rb_funcall(rstr, FORCE_ID_ENCODING, 1, UTF_8_ENCODING);
+}
+
+DEF_RV(BYTES) {
+  int32_t len = r_varint32(ss);
+  if (fld.get_lazy_type == Qnil) {
+    return rb_str_new(ss_read_chars(ss, len), len);
+  } else {
+    LazyField lf;
+    lf.ss = ss_substream(ss, len);
+    lf.fld = &fld;
+    lazy_fields->push_back(lf);
+    return Qnil;
+  }
+}
+
+DEF_RV(MESSAGE) {
+  int32_t len = r_varint32(ss);
+  ss_t tmp_ss = ss_substream(ss, len);
+  Msg& embedded_msg = *fld.embedded_msg;
+  return embedded_msg.read(tmp_ss, embedded_msg);
```

Comment:
Also, much of this code assumes that the metamessage structures are immutable and last forever (at least, until the `Pbr` instance is GCed and all the metamessages get cleaned up)